metadata/src/validation.rs

```diff
@@ -0,0 +1,851 @@
+// Copyright 2019-2023 Parity Technologies (UK) Ltd.
+// This file is dual-licensed as Apache-2.0 or GPL-3.0.
+// see LICENSE for license details.
+
+//! Utility functions for metadata validation.
+
+use frame_metadata::v15::{
+    ExtrinsicMetadata, PalletMetadata, RuntimeMetadataV15, StorageEntryMetadata, StorageEntryType,
+};
+use scale_info::{form::PortableForm, Field, PortableRegistry, TypeDef, Variant};
+use std::collections::HashSet;
+
+/// Internal byte representation for various metadata types utilized for
+/// generating deterministic hashes between different rust versions.
+#[repr(u8)]
+enum TypeBeingHashed {
+    Composite,
+    Variant,
+    Sequence,
+    Array,
+    Tuple,
+    Primitive,
+    Compact,
+    BitSequence,
+}
+
+/// Hashing function utilized internally.
+fn hash(data: &[u8]) -> [u8; 32] {
+    sp_core_hashing::twox_256(data)
+}
+
+/// XOR two hashes together. If we have two pseudorandom hashes, then this will
+/// lead to another pseudorandom value. If there is potentially some pattern to
+/// the hashes we are xoring (eg we might be xoring the same hashes a few times),
+/// prefer `hash_hashes` to give us stronger pseudorandomness guarantees.
+fn xor(a: [u8; 32], b: [u8; 32]) -> [u8; 32] {
+    let mut out = [0u8; 32];
+    for (idx, (a, b)) in a.into_iter().zip(b).enumerate() {
+        out[idx] = a ^ b;
+    }
+    out
+}
+
+/// Combine two hashes or hash-like sets of bytes together into a single hash.
+/// `xor` is OK for one-off combinations of bytes, but if we are merging
+/// potentially identical hashes, this is a safer way to ensure the result is
+/// unique.
+fn hash_hashes(a: [u8; 32], b: [u8; 32]) -> [u8; 32] {
+    let mut out = [0u8; 32 * 2];
+    for (idx, byte) in a.into_iter().chain(b).enumerate() {
```

Comment:
you also use copy_from_slice which I think is slightly cleaner/more readable here:

```rust
 out[0..32].copy_from_slice(&a[..]);
 out[32..].copy_from_slice(&b[..]);
```

Reply:
I like it! Have changed it thanks!